keras/engine/training_utils.py

```diff
@@ -71,10 +71,10 @@ def handle_partial_sample_weights(outputs, sample_weights, sample_weight_modes,
     Tuple of sample weights, one sample weight for every output, and booleans
     describing the raw sample weights.
   """
-  any_sample_weight = sample_weights is not None and any(
```

Comment:
The original logic was confusing, but the new logic does not do what's intended.

What's intended is something like

```python
if sample_weights is None:  # Most common case
  any_sample_weight = False
  partial_sample_weight = False
elif instance(sample_weights, (tuple, list, dict)):  # Multi-output case
  # Whether at least one output has sample weights
  any_sample_weight = any(w is not None for w in sample_weights)
  # Whether at least one output is missing sample weights
  partial_sample_weight = any(w is None for w in sample_weights)
```

We should go with the logic above.

Have you identified the unit tests is charge of testing this part of the code? They may need to be updated.

Reply:
There's some unit tests in keras/tests for temporal weights but I haven't found any other specific to testing sample weights. `test_fit_generator_with_sample_weight(self):` and `test_fit_generator_with_partial_sample_weight(self):` in keras/tests/temporal_sample_weights_correctness_test.py . Other than that I found tests for `Model.fit()` and `TensorLikeDataAdapter()` (which reference the relevant function) in the keras/engine folder but none of them seem to test partial weights. Is it possible that some new tests need to be written?

Regarding logic, would the logic behind this be correct? (All I did was change the `.all()` to `.any()`
`
any_sample_weight = sample_weights is not None and (
      sample_weights.numpy() != None).any()
  partial_sample_weight = any_sample_weight and (
      sample_weights.numpy() == None).any()
`

* any_sample_weight: return True when sample_weights contains None
* partial_sample_weight: return True when sample_weights contains both None and non-None.

I see that you omitted the check for non-None from the partial_sample_weight (`partial_sample_weight = any_sample_weight and` in the original code), was this intentional or a typo?  Or are you suggesting chaning how the logic is structured in the function?

Reply:
> any_sample_weight: return True when sample_weights contains None

No, it's True if sample_weights contains any non-None value. So it there's at least a sample weights array for one output.
 
> partial_sample_weight: return True when sample_weights contains both None and non-None.

Yes, it's True when some outputs have sample weights arrays and others do not.

> I see that you omitted the check for non-None from the partial_sample_weight

It should be

```
partial_sample_weight = any_sample_weight and any(w is None for w in sample_weights)
```

Can you please add a test that check the behavior for a model with 3 outputs and only some of them receive sample_weight (as a dict or list)?